torch/csrc/jit/register_special_ops.cpp

```diff
@@ -11,6 +14,126 @@ namespace torch {
 namespace jit {
 
 namespace {
+
+
+void checkListInputType(const c10::TypePtr& elem_type, const Node* node) {
+  if (!elem_type->isSubtypeOf(NumberType::get()) && elem_type != BoolType::get()) {
+    auto error = script::ErrorReport(node->getSourceLocation());
+    error << "Input list to torch.tensor must be of ints, floats, or bools, " <<
+      "got " << elem_type->str();
+    // special case empty list torch.tensor([])
+    if (elem_type->isSubtypeOf(DynamicType::get())) {
+      auto input = node->inputs().at(0);
+      if (input->node()->kind() == prim::ListConstruct && input->node()->inputs().size() == 0) {
+        error << "\n(Note: empty lists are constructed as Tensor[]; \n"
+               << "if you want an empty list of a different type, \n"
+               << "use `torch.jit.annotate(List[T], [])`, \n"
+               << "where `T` is the type of elements in the list)";
+      }
+    }
+    throw error;
+  }
+}
+
+at::ScalarType scalarTypeFromJitType(const c10::TypePtr& type) {
+  if (type == FloatType::get()) {
+    return at::ScalarType::Double;
+  } else if (type == IntType::get()) {
+    return at::ScalarType::Long;
+  } else if (type == BoolType::get()) {
+    return at::ScalarType::Byte;
+  }
+  AT_ASSERTM(0, "Add new condition, expected Float, Int, or Bool but got",
+      type->str());
+}
+
+
+int64_t list_size(IValue list) {
+  if (list.isGenericList()) {
+    return list.toGenericListRef().size();
+  } else if (list.isIntList()) {
+    return list.toIntListRef().size();
+  } else if (list.isDoubleList()){
+    return list.toDoubleListRef().size();
+  } else if (list.isBoolList()) {
+    return list.toBoolListRef().size();
+  }
+  AT_ASSERTM(0, "Unexpected list type", list);
+}
+
+std::vector<int64_t> compute_sizes(IValue seq) {
+  std::vector<int64_t> sizes;
+  // because bool, int, and float lists are specialized, inner array will
+  // will not be generic list
+  while (seq.isGenericList()) {
+    auto seq_list = seq.toGenericListRef();
+    auto length = seq_list.size();
+    AT_ASSERT(length != 0);
+    sizes.push_back(length);
+    seq = seq_list[0];
+  }
+  sizes.push_back(list_size(seq));
+  return sizes;
+}
+
+void checkSequenceSize(int64_t n, int64_t dim, int64_t seq_size) {
+  if (seq_size != n) {
+    AT_ERROR("Expected sequence of length ", n, " at dim ", dim, " (got ", seq_size, ")");
+  }
+}
+
+template <typename DTYPE>
+void storeLastDimension(char* data, const std::vector<int64_t>& sizes, const c10::ArrayRef<long>& strides, int64_t dim,
+    int elementSize, std::vector<DTYPE> obj) {
+  auto n = sizes[dim];
+  auto seq_size = obj.size();
+  checkSequenceSize(n, dim, seq_size);
+  for (int64_t i = 0; i < n; i++) {
+    *(DTYPE*)data = obj[i];
+    data += strides[dim] * elementSize;
+  }
+}
+
+// bool vector needs to be cast to uint8_t
+template<>
+void storeLastDimension<bool>(char* data, const std::vector<int64_t>& sizes, const c10::ArrayRef<long>& strides, int64_t dim,
```

Comment:
Why is this specialization necessary? Shouldn't it be fine since `char` and `uint8_t` are defined to both be 1 byte?

Reply:
This line:
```
*(uint8_t*)data = static_cast<uint8_t>(obj[i]);
```

There needs to be the cast from bool to uint8. From what I googled / saw on stack overflow, static_cast on the same type isn't guaranteed to be a noop, plus having extra casts when they're not needed seems like misleading code.